### PR TITLE
Improve log messages on VolumeOrchestrator class

### PR DESCRIPTION
**Pull request by paulazomig**
### Description

This PR aims to improve several log entries in the VolumeOrchestrator class by adding more clear and precise information to facilitate future troubleshooting.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
The changes were tested by checking the log entries during volume manipulation (creating, deleting, migrating volumes between different storages, creating volumes from snapshots, etc.).

### CODE REVIEW

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 1]
```diff
@@ -484,13 +489,14 @@ public VolumeInfo createVolumeFromSnapshot(Volume volume, Snapshot snapshot, Use
             }
 
             if (pool == null) {
+                String logMsg = String.format("Could not find a storage pool in the pod/cluster of the provided VM [%s] to create the volume [%s] in.", vm, volumeToString);
+
                 //pool could not be found in the VM's pod/cluster.
                 if (s_logger.isDebugEnabled()) {
-                    s_logger.debug("Could not find any storage pool to create Volume in the pod/cluster of the provided VM " + vm.getUuid());
+                    s_logger.error(logMsg);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:496]: remove the if (isDebugEnabled()). doesn´t make  sense anymore.

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 2]
```diff
@@ -546,15 +554,18 @@ public VolumeInfo createVolumeFromSnapshot(Volume volume, Snapshot snapshot, Use
...
         AsyncCallFuture<VolumeApiResult> future = volService.createVolumeFromSnapshot(vol, store, snapInfo);
+        String snapshotToString = getReflectOnlySelectedFields(snapInfo.getSnapshotVO());
+
         try {
             VolumeApiResult result = future.get();
             if (result.isFailed()) {
-                s_logger.debug("Failed to create volume from snapshot:" + result.getResult());
-                throw new CloudRuntimeException("Failed to create volume from snapshot:" + result.getResult());
+                String logMsg = String.format("Failed to create volume from snapshot [%s] due to [%s].", snapshotToString, result.getResult());
+                s_logger.error(logMsg);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:563]: :+1:

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java [hunk 3]
```diff
@@ -91,6 +91,9 @@ public class PrimaryDataStoreImpl implements PrimaryDataStore {
     private VolumeDao volumeDao;
     private Map<String, String> _details;
 
+    private String uuid;
```
Comment (DaanHoogland) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java:94]: why the changes in this file? they seem to only add maintenance risk.

File: engine/api/src/main/java/org/apache/cloudstack/engine/orchestration/service/VolumeOrchestrationService.java [hunk 4]
```diff
@@ -112,7 +112,7 @@ DiskProfile allocateRawVolume(Type type, String name, DiskOffering offering, Lon
 
     void release(VirtualMachineProfile profile);
 
-    void release(long vmId, long hostId);
+    void release(VirtualMachine vm, Host host);
```
Comment (DaanHoogland) [inline engine/api/src/main/java/org/apache/cloudstack/engine/orchestration/service/VolumeOrchestrationService.java:115]: I am not to sure about passing objects from service to service here. Is this change needed?

**Review by DaanHoogland — COMMENTED** [review]
I like the premise, but some of the implementation details worry me.

**Comment by DaanHoogland** [comment]
> I like the premise, but some of the implementation details worry me.

btw, I reviewed the sonarcloud bugs reported. I think we can ignore these for now as these are not really new ones.
As for the code smells, please have a look. Some of those can certainly be addressed;

> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=apache_cloudstack&pullRequest=6408)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=BUG) [C](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=BUG) [6 Bugs](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6408&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6408&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=6408&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=CODE_SMELL) [10 Code Smells](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=6408&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6408&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6408&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6408&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=6408&metric=new_duplicated_lines_density&view=list)

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java (hunk 3 above)
Comment (paulazomig) [reply engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java:94]: @DaanHoogland the changes here were made in order to have the proper information to identify the objects in the log messages

File: engine/api/src/main/java/org/apache/cloudstack/engine/orchestration/service/VolumeOrchestrationService.java (hunk 4 above)
Comment (paulazomig) [reply engine/api/src/main/java/org/apache/cloudstack/engine/orchestration/service/VolumeOrchestrationService.java]: @DaanHoogland removed the changes

**Comment by paulazomig** [comment]
@DaanHoogland About the code smells, most of them were not related to the changes made in this PR, but I addressed the ones that were and could be improved.

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 5]
```diff
@@ -638,46 +655,57 @@ public VolumeInfo createVolume(VolumeInfo volume, VirtualMachine vm, VirtualMach
...
-            s_logger.debug("Trying to create " + volume + " on " + pool);
+            s_logger.debug(String.format("Trying to create volume [%s] on storage pool [%s].",
+                    volumeToString, poolToString));
         }
         DataStore store = dataStoreMgr.getDataStore(pool.getId(), DataStoreRole.Primary);
         for (int i = 0; i < 2; i++) {
             // retry one more time in case of template reload is required for Vmware case
             AsyncCallFuture<VolumeApiResult> future = null;
-            boolean isNotCreatedFromTemplate = volume.getTemplateId() == null ? true : false;
+            boolean isNotCreatedFromTemplate = volumeInfo.getTemplateId() == null ? true : false;
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:673]:
```suggestion
            boolean isNotCreatedFromTemplate = (volumeInfo.getTemplateId() == null);
```

Comment (DaanHoogland) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:673]: not due to your code @paulazomig , but an easy improvement.

Comment (DaanHoogland) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:673]: there are some others, like reducing cyclic complexity and extracting a string constant that would be easy wins. Not a -1, but easy wins they would be.

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java:94]: The extra fields are not exposed and the `psdv` is available to retrieve the `uuid` and `name`, so why would you store the fields extra? the old implementation of `getUuid()` and `getName()` seems to suffice.

Comment (paulazomig) [reply engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java:94]: @DaanHoogland, sorry, my explanation was not clear in the first comment.

Those fields were added this way because we're using `ReflectionToStringBuilderUtils` to build the string to print in the log. As `ReflectionToStringBuilderUtils` uses the object attributes to build the string, this object is declared as an interface `StoragePool` (in runtime this object is an instance of `PrimaryDataStoreImpl`) and we cannot retrieve the `StoragePoolVO` contained in it, we declared these attributes (that already are there) to be reflected later.

Comment (DaanHoogland) [reply engine/storage/volume/src/main/java/org/apache/cloudstack/storage/datastore/PrimaryDataStoreImpl.java:94]: ok, acceptable ;)

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 6]
```diff
@@ -1831,7 +1917,7 @@ public void cleanupStorageJobs() {
                     cleanupVolumeDuringSnapshotFailure(work.getVolumeId(), work.getSnapshotId());
                 }
             } catch (Exception e) {
-                s_logger.debug("clean up job failure, will continue", e);
+                s_logger.error(String.format("Clean up job failed due to [%s]. Will continue with other clean up jobs.", e.getMessage()), e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:1920]: please do not log an exception on non fatal errors. if needed add a debug statement that logs the stacktrace but keep info and up clean.

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 7]
```diff
@@ -1250,18 +1293,18 @@ public Volume liveMigrateVolume(Volume volume, StoragePool destPool) {
...
                 return null;
             }
             return result.getVolume();
-        } catch (InterruptedException e) {
-            s_logger.debug("migrate volume failed", e);
-            return null;
-        } catch (ExecutionException e) {
-            s_logger.debug("migrate volume failed", e);
+        } catch (InterruptedException | ExecutionException e) {
+            s_logger.error(String.format("Volume [%s] migration failed due to [%s].", volToString, e.getMessage()), e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:1307]: please don´t log error with exception. If need be add a debug that logs the stacktrace

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 8]
```diff
@@ -1235,12 +1280,10 @@ public Volume migrateVolume(Volume volume, StoragePool destPool) throws StorageU
...
             return result.getVolume();
-        } catch (InterruptedException e) {
-            s_logger.debug("migrate volume failed", e);
-            throw new CloudRuntimeException(e.getMessage());
-        } catch (ExecutionException e) {
-            s_logger.debug("migrate volume failed", e);
-            throw new CloudRuntimeException(e.getMessage());
+        } catch (InterruptedException | ExecutionException e) {
+            String msg = String.format("Volume [%s] migration failed due to [%s].", volumeToString, e.getMessage());
+            s_logger.error(msg, e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:1285]: please don´t log error with exception. If need be add a debug that logs the stacktrace

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 9]
```diff
@@ -1153,13 +1192,13 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
...
+
             future = volService.expungeVolumeAsync(volFactory.getVolume(expunge.getId()));
             try {
                 future.get();
-            } catch (InterruptedException e) {
-                s_logger.debug("failed expunge volume" + expunge.getId(), e);
-            } catch (ExecutionException e) {
-                s_logger.debug("failed expunge volume" + expunge.getId(), e);
+            } catch (InterruptedException | ExecutionException e) {
+                s_logger.error(String.format("Failed to expunge volume [%s] due to [%s].", expungeToString, e.getMessage()), e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:1201]: please don´t log error with exception. If need be add a debug that logs the stacktrace

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 10]
```diff
@@ -1061,16 +1096,16 @@ public VolumeVO doInTransaction(TransactionStatus status) {
...
                 // In case of VMware VM will continue to use the old root disk until expunged, so force expunge old root disk
                 if (vm.getHypervisorType() == HypervisorType.VMware) {
-                    s_logger.info("Expunging volume " + existingVolume.getId() + " from primary data store");
+                    s_logger.info(String.format("Trying to expunge volume [%s] from primary data storage.", volumeToString));
                     AsyncCallFuture<VolumeApiResult> future = volService.expungeVolumeAsync(volFactory.getVolume(existingVolume.getId()));
                     try {
                         future.get();
                     } catch (Exception e) {
-                        s_logger.debug("Failed to expunge volume:" + existingVolume.getId(), e);
+                        s_logger.error(String.format("Failed to expunge volume [%s] from primary data storage due to [%s].", volumeToString, e.getMessage()), e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:1108]: please don´t log error with exception. If need be add a debug that logs the stacktrace

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 11]
```diff
@@ -638,46 +655,57 @@ public VolumeInfo createVolume(VolumeInfo volume, VirtualMachine vm, VirtualMach
...
                 return result.getVolume();
-            } catch (InterruptedException e) {
-                s_logger.error("create volume failed", e);
-                throw new CloudRuntimeException("create volume failed", e);
-            } catch (ExecutionException e) {
-                s_logger.error("create volume failed", e);
-                throw new CloudRuntimeException("create volume failed", e);
+            } catch (InterruptedException | ExecutionException e) {
+                String msg = String.format("Failed to create volume [%s] due to [%s].", volumeToString, e.getMessage());
+                s_logger.error(msg, e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:696]: please don´t log error with exception. If need be add a debug that logs the stacktrace

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 12]
```diff
@@ -546,15 +553,18 @@ public VolumeInfo createVolumeFromSnapshot(Volume volume, Snapshot snapshot, Use
...
-                throw new CloudRuntimeException("Failed to create volume from snapshot:" + result.getResult());
+                String logMsg = String.format("Failed to create volume from snapshot [%s] due to [%s].", snapshotToString, result.getResult());
+                s_logger.error(logMsg);
+                throw new CloudRuntimeException(logMsg);
             }
             return result.getVolume();
         } catch (InterruptedException | ExecutionException e) {
-            String message = String.format("Failed to create volume from snapshot [%s] due to [%s].", snapInfo.getTO(), e.getMessage());
+            String message = String.format("Failed to create volume from snapshot [%s] due to [%s].", snapshotToString, e.getMessage());
             s_logger.error(message, e);
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:568]: not your doing, but please don´t log error with exception. If need be add a debug that logs the stacktrace

**Review by GutoVeronezi — APPROVED** [review]
CLGTM
